quic/s2n-quic-core/src/state.rs

```diff
@@ -0,0 +1,210 @@
+        pub fn dot() -> impl ::core::fmt::Display {
+            struct Dot;
+
+            impl ::core::fmt::Display for Dot {
+                fn fmt(&self, f: &mut ::core::fmt::Formatter) -> ::core::fmt::Result {
+                    writeln!(f, "digraph {{")?;
+
+                    let mut all_states = [
+                        // collect all of the states we've observed
+                        $(
+                            $(
+                                stringify!($valid),
+                            )*
+                            stringify!($target),
+                        )*
+                    ];
+
+                    let all_states = $crate::state::dedup_states(&mut all_states);
+
+                    for state in all_states {
+                        writeln!(f, "  {state};")?;
+                    }
+
+                    $(
+                        $(
+                            writeln!(
+                                f,
+                                "  {} -> {} [label = {:?}];",
+                                stringify!($valid),
+                                stringify!($target),
+                                stringify!($event),
+                            )?;
+                        )*
+                    )*
+
+                    writeln!(f, "}}")?;
+                    Ok(())
+                }
+            }
+
+            Dot
+        }
+    }
+}
+
+pub use crate::__state_event__ as event;
+
+#[doc(hidden)]
+pub fn dedup_states<'a>(states: &'a mut [&'a str]) -> &'a mut [&'a str] {
+    // TODO replace with
+    // https://doc.rust-lang.org/std/primitive.slice.html#method.partition_dedup
+    // when stable
+    //
+    // For now, we've just inlined their implementation
+
+    let len = states.len();
+    if len <= 1 {
+        return states;
+    }
+
+    // sort the states before deduping
+    states.sort_unstable();
+
+    let ptr = states.as_mut_ptr();
+    let mut next_read: usize = 1;
+    let mut next_write: usize = 1;
+
+    // SAFETY: the `while` condition guarantees `next_read` and `next_write`
+    // are less than `len`, thus are inside `self`. `prev_ptr_write` points to
+    // one element before `ptr_write`, but `next_write` starts at 1, so
+    // `prev_ptr_write` is never less than 0 and is inside the slice.
+    // This fulfils the requirements for dereferencing `ptr_read`, `prev_ptr_write`
+    // and `ptr_write`, and for using `ptr.add(next_read)`, `ptr.add(next_write - 1)`
+    // and `prev_ptr_write.offset(1)`.
+    //
+    // `next_write` is also incremented at most once per loop at most meaning
+    // no element is skipped when it may need to be swapped.
+    //
+    // `ptr_read` and `prev_ptr_write` never point to the same element. This
+    // is required for `&mut *ptr_read`, `&mut *prev_ptr_write` to be safe.
+    // The explanation is simply that `next_read >= next_write` is always true,
+    // thus `next_read > next_write - 1` is too.
```

Comment:
nit: not sure its worth having this big scary `SAFETY` comment here if its just inlined from rust, maybe just link to the comment in the source

Reply:
Yeah I debated to include it or not. But I did want to minimize the diff from the rust impl (which is where this came from) so i ended up keeping it.